metabase/metabase
# Graph gymnastics to handle type coercing and promotion

### Description

This follows on from the [int to float promotion](https://github.com/metabase/metabase/pull/39493) work, to keep float-to-int coercion working as well.

It turns out to be quite complex, but also served as a nice stress test of the `ordered-hierarchy`, and it was gratifying to see it handle it.

Two things that occur to me are:

1. Perhaps this complexity isn't worth it, and we should just let these `::float-or-int` values force a column promotion.
2. Getting this behavior might have been easier if we relaxed our DAG to a graph, instead of inserting two more abstract nodes to avoid a cycle. I'm not sure how those details would work out, but I have a hunch there's a way. Might just add a whole different kettle of complex lifeforms though.

### How to verify

See that the test suite is passing again - specifically this test:

```clojure
{:upload-type ::upload/int, :uncoerced "2.0", :coerced 2} ;; upload-test/append-type-coercion-test
```
